ckan/views/api.py

```diff
@@ -95,17 +105,413 @@ def _finish_ok(response_data=None,
     return _finish(status_int, response_data, content_type, headers)
 
 
+def _finish_not_authz(extra_msg=None):
+    response_data = _(u'Access denied')
+    if extra_msg:
+        response_data = u'%s - %s' % (response_data, extra_msg)
+    return _finish(403, response_data, u'json')
+
+
+def _finish_not_found(extra_msg=None):
+    response_data = _(u'Not found')
+    if extra_msg:
+        response_data = u'%s - %s' % (response_data, extra_msg)
+    return _finish(404, response_data, u'json')
+
+
+def _finish_bad_request(extra_msg=None):
+    response_data = _(u'Bad request')
+    if extra_msg:
+        response_data = u'%s - %s' % (response_data, extra_msg)
+    return _finish(400, response_data, u'json')
+
+
 def _wrap_jsonp(callback, response_msg):
     return u'{0}({1});'.format(callback, response_msg)
 
 
+def _get_request_data(try_url_params=False):
+    u'''Returns a dictionary, extracted from a request.
+
+    If there is no data, None or "" is returned.
+    ValueError will be raised if the data is not a JSON-formatted dict.
+
+    The data is retrieved as a JSON-encoded dictionary from the request
+    body.  Or, if the `try_url_params` argument is True and the request is
+    a GET request, then an attempt is made to read the data from the url
+    parameters of the request.
+
+    try_url_params
+        If try_url_params is False, then the data_dict is read from the
+        request body.
+
+        If try_url_params is True and the request is a GET request then the
+        data is read from the url parameters.  The resulting dict will only
+        be 1 level deep, with the url-param fields being the keys.  If a
+        single key has more than one value specified, then the value will
+        be a list of strings, otherwise just a string.
+
+    '''
+    def make_unicode(entity):
+        u'''Cast bare strings and strings in lists or dicts to Unicode. '''
+        if isinstance(entity, str):
+            return unicode(entity)
+        elif isinstance(entity, list):
+            new_items = []
+            for item in entity:
+                new_items.append(make_unicode(item))
+            return new_items
+        elif isinstance(entity, dict):
+            new_dict = {}
+            for key, val in entity.items():
+                new_dict[key] = make_unicode(val)
+            return new_dict
+        else:
+            return entity
```

Comment:
This kind of recursive function is really slow over long lists and deeply nested data, why is it required?

Reply:
I ported this straight from the current API controller, as many did before me starting from this ancient commit 8c4d651addab960ffd50ec50deab7d5014abf6f2 in 2010 that seem to be related to the REST API.

I'm pretty sure that Flask [will take care](http://flask.pocoo.org/docs/0.12/unicode/) of all the unicode handle automatically so I'll just remove it.